### PR TITLE
feat(ssr): enhance support for Universal and SSR with stylesheets

### LINKED ISSUE #373: Universal pre-rendered styles are not correct
With last nightly build, there are no more blocking errors with Flex Layout in Universal, but the pre-rendered styles are not correct.

Repro : 
- `git clone https://github.com/cyrilletuzi/nguniversal-expressengine-httpinterceptor.git`
- `git checkout -b flexlayoutissue`
- `npm install`
- `npm start`
- go to localhost:3000 : rendered correctly with JavaScript (vertical in small screens, horizontal in desktop screens with 3 stretched items by row)
- disable JavaScript to test what the server is pre-rendering and refresh : not rendered correctly (horizontal with the 4 items on the same line, on all screen sizes)

The only flex relative code is in `src/app/app.component` (the other files are just Universal boilerplate, and `innerHTML` is just to quickly test different heights) : 

```html
<div fxLayout.xs="column" fxLayoutWrap.xs="nowrap">
  <div class="test" fxFlex.gt-xs="calc(100% / 3)" *ngFor="let item of list" [innerHTML]="item"></div>
</div>
```

##### Mobile Viewport 

screen shot 2017-08-09 at 10 53 16 am

##### Desktop Viewport

screen shot 2017-08-09 at 10 24 07 am

----

##### Mobile Viewport 

screen shot 2017-08-09 at 10 53 01 am

##### Desktop Viewport

screen shot 2017-08-09 at 10 29 35 am

### PR DESCRIPTION
**Pull request by CaerusKaru**
* Add `StyleService` class to manage application and retrieval of styles from elements in a platform-agnostic manner
* Add virtual stylesheet to store server styles, which applies default styles when breakpoint overrides are not present
* While not in the browser (ssr), intercept all style calls and reroute them to the virtual stylesheet
* For server-side rendering, add a new type of MediaQueryList similar to the MockMediaQueryList to support manual activation/deactivation of breakpoints
* Add jasmine testing mode for SSR
* Add FlexLayoutServerModule to invoke SSR styling
* Remove unnecessary Renderer references and replace them with DOM APIs
* Add whitespace debugging mode for server styles

Fixes #373.

> See [Design Doc](https://docs.google.com/document/d/1fg04ihw42dJJHGd6fugdiBe39iJot8aErhiE7CjwfmQ/edit#)

### CODE REVIEW

**Comment by CaerusKaru** [comment]
Example output for the following HTML markup:

```html
<div class="night-theme">
  <div fxLayout="row" fxLayout.xs="column" style="height:500px" ngxSplit="row">
    <div fxFlex="30%" ngxSplitArea class="c1r1">
      <div class="c1r1_header">Column #1 - Row #1</div>
      <ul>
        <li>2 Columns: 30% + 70%</li>
        <li>2nd Column: 2 rows</li>
        <li>2nd Column: 50% + 50%</li>
      </ul>
    </div>
    <div class="handle handle-row" ngxSplitHandle>
      <i class="material-icons">&#xE25D;</i>
    </div>
    <div fxFlex.xs="70%" fxFlex.gt-md="50%" fxFlex.lg="60%" ngxSplitArea>
      <div fxLayout="column" fxFlexFill ngxSplit="column">
        <div fxFlex="50%" ngxSplitArea class="c2r1_body">
          <div class="c2r1_header">Column #2 - Row #1</div>
          <h1>Layout Dashboard</h1>
          <p>
            Demonstrate use of ngxSplit with the Flex-Layout API
            and flexbox css layouts.
            <br/><br/>
            Haxx0r ipsum cd ctl-c Starcraft concurrently salt unix baz class bar linux
            January 1, 1970 syn for mutex daemon todo mountain dew recursively. Mainframe
            wannabee machine code hack the mainframe do void python bin big-endian break
            tcp ddos emacs public frack.Over clock headers data private *.* pwned
            fork script kiddies.
          </p>
        </div>
        <div class="handle handle-column" ngxSplitHandle>
          <i class="material-icons">&#xE25D;</i>
        </div>
        <div fxFlex="50%" ngxSplitArea class="c2r2">
          <div class="c2r2_header">Column #2 - Row #2</div>
          <ul>
            <li>List Item #1</li>
            <li>List Item #2</li>
            <li>List Item #3</li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
```

```css
@media all {
     .flex-layout-0 {
         height: 100%;
         min-height: 100%;
         min-width: 100%;
         width: 100%;
         -webkit-flex-direction: column;
         box-sizing: border-box;
         display: flex;
         flex-direction: column;
    }
     .flex-layout-1 {
         -webkit-flex-direction: row;
         box-sizing: border-box;
         display: flex;
         flex-direction: row;
    }
     .flex-layout-2 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
         max-width: 30%;
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
    }
     .flex-layout-3 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
    }
     .flex-layout-4 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
         max-height: 50%;
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
    }
     .flex-layout-5 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
         max-height: 50%;
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
    }
}
 @media (min-width: 1280px) and (max-width: 1919px) {
     .flex-layout-1 {
         -webkit-flex-direction: row;
         box-sizing: border-box;
         display: flex;
         flex-direction: row;
    }
     .flex-layout-3 {
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         box-sizing: border-box;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
         max-width: 60%;
    }
}
 @media (max-width: 1279px) {
     .flex-layout-1 {
         -webkit-flex-direction: row;
         box-sizing: border-box;
         display: flex;
         flex-direction: row;
    }
     .flex-layout-3 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
    }
}
 @media (min-width: 1280px) {
     .flex-layout-1 {
         -webkit-flex-direction: row;
         box-sizing: border-box;
         display: flex;
         flex-direction: row;
    }
     .flex-layout-3 {
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         box-sizing: border-box;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
         max-width: 50%;
    }
}
 @media (min-width: 960px) and (max-width: 1279px) {
     .flex-layout-1 {
         -webkit-flex-direction: row;
         box-sizing: border-box;
         display: flex;
         flex-direction: row;
    }
     .flex-layout-3 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
    }
}
 @media (min-width: 0px) and (max-width: 599px) {
     .flex-layout-1 {
         -webkit-flex-direction: column;
         box-sizing: border-box;
         display: flex;
         flex-direction: column;
    }
     .flex-layout-2 {
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         box-sizing: border-box;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
         max-height: 30%;
    }
     .flex-layout-3 {
         -webkit-flex: 1 1 0.000000001px;
         box-sizing: border-box;
         flex: 1 1 0.000000001px;
         max-height: 70%;
         -webkit-flex-basis: 100%;
         -webkit-flex-grow: 1;
         -webkit-flex-shrink: 1;
         flex-basis: 100%;
         flex-grow: 1;
         flex-shrink: 1;
    }
}
```

**Comment by CaerusKaru** [comment]
Note: this needs a testing mode where we can change the PLATFORM_ID flag from browser mode to server mode and confirm all tests still pass.

**Comment by naveedahmed1** [comment]
@CaerusKaru should I create a new issue to track the combining of classes for the common styles?

**Comment by CaerusKaru** [comment]
@naveedahmed1 You can add it to #373

**Comment by CaerusKaru** [comment]
@jelbourn @vikerman

File: src/lib/api/ext/img-src.ts [hunk 1]
```diff
@@ -57,12 +60,16 @@ export class ImgSrcDirective extends BaseFxDirective implements OnInit, OnChange
...
-    super(monitor, elRef, renderer, platformId);
-    this._cacheInput('src', elRef.nativeElement.getAttribute('src') || '');
+  constructor(protected _elRef: ElementRef,
+              protected _monitor: MediaMonitor,
+              protected _styler: StyleService,
+              @Inject(PLATFORM_ID) protected _platformId: Object,
+              @Optional() @Inject(SERVER_TOKEN) protected _serverModuleLoaded: boolean) {
+    super(_monitor, _elRef, _styler);
+    this._cacheInput('src', _elRef.nativeElement.getAttribute('src') || '');
+    if (isPlatformServer(this._platformId) && this._serverModuleLoaded) {
```
Comment (vikerman) [inline src/lib/api/ext/img-src.ts]: Can you explain a bit on how the serverModuleLoaded flag is used?

**Comment by vikerman** [comment]
The platform-server part and associated tests looks good to me(I didn't go through the actual styles logic). 

Just need a small clarification.

(In the future it would help if the PRs are collection of smaller PRs/commits)

File: src/lib/api/ext/img-src.ts (hunk 1 above)
Comment (CaerusKaru) [reply src/lib/api/ext/img-src.ts]: The flag is to determine if the `FlexLayoutServerModule` is included (since it's the only module that provides that token).

Loading the server module is optional -- if a user doesn't load it, it means they want the inline styles as a fallback (it also prevents this from being a breaking change, since it's totally opt-in).

Comment (vikerman) [reply src/lib/api/ext/img-src.ts]: Will this get into the docs?

Comment (CaerusKaru) [reply src/lib/api/ext/img-src.ts]: Absolutely. Docs will be in a follow-up PR once this has all been finalized

File: src/lib/media-query/match-media.ts [hunk 2]
```diff
@@ -5,42 +5,83 @@
...
+    if (this._listeners.indexOf(listener) === -1) {
+      this._listeners.push(listener);
+    }
+    if (this._isActive) {
+      listener(this);
+    }
+  }
+
+  removeListener(_: MediaQueryListListener) {
+  }
```
Comment (jelbourn) [inline src/lib/media-query/match-media.ts]: Add comment in function body here like
```ts
// Don't need to remove listeners in the server environment.
```

File: src/lib/media-query/match-media.ts [hunk 3]
```diff
@@ -5,42 +5,83 @@
...
- * EventHandler callback with the mediaQuery [range] activates or deactivates
+ * Special server-only class to simulate a MediaQueryList and
+ * - supports manual activation to simulate mediaQuery matching
+ * - manages listeners
  */
-export interface MediaQueryListListener {
-  // Function with Window's MediaQueryList argument
-  (mql: MediaQueryList): void;
-}
+export class ServerMediaQueryList implements MediaQueryList {
```
Comment (jelbourn) [inline src/lib/media-query/match-media.ts]: This should be in a separate file
(generally all server-specific stuff should be in separate files so that you can better ensure you're only getting what you need on the client)

File: src/lib/media-query/match-media.ts [hunk 4]
```diff
@@ -51,19 +92,36 @@ export interface MediaQueryList {
  */
 @Injectable()
 export class MatchMedia {
-  protected _registry: Map<string, MediaQueryList>;
+  protected _registry: Map<string, MediaQueryList|ServerMediaQueryList>;
```
Comment (jelbourn) [inline src/lib/media-query/match-media.ts]: Instead of using a union type, it would be better to use an interface and an `InjectionToken`. When using a union type, you may end up getting the server-side code on the client

File: src/lib/server/server-provider.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+let UNIQUE_CLASS = 0;
+const DEBUG_FLAG = false;
+
+/**
+ * create @media queries based on a virtual stylesheet
+ * * Adds a unique class to each element and stores it
+ *   in a shared classMap for later reuse
+ */
+function formatStyle(stylesheet: Map<HTMLElement, Map<string, string|number>>,
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Something like `generateCss`?

File: src/lib/server/server-provider.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+                     _document: Document,
+                     mediaQuery: string,
+                     classMap: Map<HTMLElement, string>) {
+  let styleText = DEBUG_FLAG ? `
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Rather than formatting the css manually, it would probably be better to use some existing node css formatter / prettifier (still only when debugging). As it is now, the method is pretty hard to mentally parse.

File: src/lib/server/server-provider.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  BreakPoint,
+  BREAKPOINTS,
+  CLASS_NAME,
+  MatchMedia,
+  ServerStylesheet,
+  SERVER_TOKEN,
+} from '@angular/flex-layout';
+
+let UNIQUE_CLASS = 0;
+const DEBUG_FLAG = false;
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: `IS_DEBUG_MODE`?

File: src/lib/server/server-provider.ts (hunk 5 above)
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Add JsDoc parameter descriptions

File: src/lib/server/server-provider.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ * platform
+ * format the static @media queries for all breakpoints
+ * to be used on the server and append them to the <head>
+ */
+export function addStyles(serverSheet: ServerStylesheet,
+                          matchMedia: MatchMedia,
+                          _document: Document,
+                          breakpoints: BreakPoint[]) {
+  // necessary because of angular/angular/issues/14485
+  const res = () => {
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: `res` isn't a very meaningful variable name

File: src/lib/server/server-provider.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  return styleText;
+}
+
+/**
+ * Add or remove static styles depending on the current
+ * platform
+ * format the static @media queries for all breakpoints
+ * to be used on the server and append them to the <head>
+ */
+export function addStyles(serverSheet: ServerStylesheet,
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Add JsDoc param descriptions

File: src/lib/server/server-provider.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                          matchMedia: MatchMedia,
+                          _document: Document,
+                          breakpoints: BreakPoint[]) {
+  // necessary because of angular/angular/issues/14485
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: I don't see how this issue is related

File: src/lib/server/server-provider.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  const res = () => {
+    const styleTag = _document.createElement('style');
+    const classMap = new Map<HTMLElement, string>();
+    const defaultStyles = new Map(serverSheet.stylesheet);
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Add descriptions for these variables?

File: src/lib/server/server-provider.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    let styleText = formatStyle(defaultStyles, _document, 'all', classMap);
+
+    breakpoints.reverse();
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: Add comment for what this block of code is doing?

File: src/lib/server/server-provider.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  return res;
+}
+
+/**
+ *  Provider to set static styles on the server
+ */
+export const SERVER_PROVIDERS = [
+  {
+    provide: BEFORE_APP_SERIALIZED,
+    useFactory: addStyles,
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: I would create a separate thing like `FLEX_SSR_SERIALIZER_FACTORY` that then calls the more specific functions

File: src/lib/server/server-provider.ts (hunk 9 above)
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: `addStyles` is not very specific. Could the name be something more like `generateStaticFlexLayoutStyles`? It would also be good if the description went into more detail about how it works.

File: src/lib/utils/styling/browser-provider.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+ */
+import {
+  APP_BOOTSTRAP_LISTENER,
+  PLATFORM_ID,
+  InjectionToken,    // tslint:disable-line:no-unused-variable
+  ComponentRef,      // tslint:disable-line:no-unused-variable
+} from '@angular/core';
+import {DOCUMENT, isPlatformBrowser} from '@angular/common';
+
+export function removeStyles(_document: Document, platformId: Object) {
```
Comment (jelbourn) [inline src/lib/utils/styling/browser-provider.ts]: Description?

File: src/lib/utils/styling/browser-provider.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {
+  APP_BOOTSTRAP_LISTENER,
+  PLATFORM_ID,
+  InjectionToken,    // tslint:disable-line:no-unused-variable
+  ComponentRef,      // tslint:disable-line:no-unused-variable
```
Comment (jelbourn) [inline src/lib/utils/styling/browser-provider.ts]: Why do these unused symbols need to stay here?

File: src/lib/utils/styling/browser-provider.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+} from '@angular/core';
+import {DOCUMENT, isPlatformBrowser} from '@angular/common';
+
+export function removeStyles(_document: Document, platformId: Object) {
+  // necessary because of angular/angular/issues/14485
```
Comment (jelbourn) [inline src/lib/utils/styling/browser-provider.ts]: How is this issue related?

File: src/lib/utils/styling/server-stylesheet.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {Injectable} from '@angular/core';
+
+@Injectable()
+export class ServerStylesheet {
```
Comment (jelbourn) [inline src/lib/utils/styling/server-stylesheet.ts]: Class and method descriptions?

File: src/lib/utils/styling/server-token.ts [hunk 18]
```diff
@@ -0,0 +1,10 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {InjectionToken} from '@angular/core';
+
+export const SERVER_TOKEN = new InjectionToken<boolean>('FlexLayoutServerLoaded');
```
Comment (jelbourn) [inline src/lib/utils/styling/server-token.ts]: Description?

File: src/lib/utils/styling/styler.ts [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+ */
+import {Inject, Injectable, Optional, PLATFORM_ID} from '@angular/core';
+import {isPlatformBrowser, isPlatformServer} from '@angular/common';
+
+import {applyCssPrefixes} from '../auto-prefixer';
+import {ServerStylesheet} from './server-stylesheet';
+import {SERVER_TOKEN} from './server-token';
+
+@Injectable()
+export class StyleService {
```
Comment (jelbourn) [inline src/lib/utils/styling/styler.ts]: Class description?

On naming: my rule is to avoid the term "Service" in class names because it doesn't tell you anything about what the class actually does. You'll notice in material we have things like `ScrollDispatcher`, `FocusMonitor`, `InteractivityChecker`, etc. instead of `ScrollService`, `FocusService`, `InteracitivityService`. Adding a more specific noun makes the name much more meaningful.

File: src/lib/utils/styling/styler.ts [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+  /**
+   * Find the DOM element's inline style value (if any)
+   */
+  lookupInlineStyle(element: HTMLElement, styleName: string): string {
+    return element.style[styleName] || element.style.getPropertyValue(styleName);
+  }
+
+  /**
+   * Determine the inline or inherited CSS style
+   * @TODO(CaerusKaru): platform-server has no implementation for getComputedStyle
```
Comment (jelbourn) [inline src/lib/utils/styling/styler.ts]: What's the action on this TODO? The whole idea of `getComputedStyle` inherently can't work on the server (without being driven by an entire browser rendering engine)

File: src/lib/utils/styling/styler.ts [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    let value = '';
+    if (element) {
+      let immediateValue = value = this.lookupInlineStyle(element, styleName);
+      if (!immediateValue) {
+        if (isPlatformBrowser(this._platformId)) {
+          if (!inlineOnly) {
+            value = getComputedStyle(element).getPropertyValue(styleName);
+          }
+        } else {
+          if (this._serverModuleLoaded) {
```
Comment (jelbourn) [inline src/lib/utils/styling/styler.ts]: There are a lot of places throughout the PR where you're explicitly performing a check for browser vs. server and performing a different action for each. In general it would be better to delegate to some injectable with a common interface that uses a different implementation on each platform.

File: src/lib/utils/styling/styler.ts [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+      }
+    });
+  }
+}
+
+/**
+ * Definition of a css style. Either a property name (e.g. "flex-basis") or an object
+ * map of property name and value (e.g. {display: 'none', flex-order: 5})
+ */
+export type StyleDefinition = string | { [property: string]: string | number | null };
```
Comment (jelbourn) [inline src/lib/utils/styling/styler.ts]: Why do either an object literal or string instead of just always doing an object literal?

File: src/lib/utils/styling/browser-provider.ts (hunk 15 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/browser-provider.ts]: Because otherwise we get the following TS error: 

```
error TS4023: Exported variable 'MATCH_MEDIA_PROVIDER' has or is using name 'InjectionToken' from external module "/Users/main/flex-layout/node_modules/@angular/core/src/di/injection_token" but cannot be named
```

File: src/lib/utils/testing/custom-matchers.ts [hunk 23]
```diff
@@ -204,7 +205,7 @@ export const customMatchers: jasmine.CustomMatcherFactories = {
  * specified DOM element.
  */
 function buildCompareStyleFunction(inlineOnly = true) {
-  return function (actual: any, styles: { [k: string]: string } | string) {
+  return function (actual: any, styles: { [k: string]: string } | string, styler: StyleService) {
```
[the remark this answers is not in the document]
Comment (CaerusKaru) [reply src/lib/utils/testing/custom-matchers.ts]: Discussed offline, unfortunately very necessary :-1:

File: src/lib/media-query/match-media.ts (hunk 2 above)
Comment (CaerusKaru) [reply src/lib/media-query/match-media.ts]: Added

File: src/lib/media-query/match-media.ts (hunk 3 above)
Comment (CaerusKaru) [reply src/lib/media-query/match-media.ts]: Added to a separate file with a separate implementation of `MatchMedia` for the server as suggested below

File: src/lib/media-query/match-media.ts (hunk 4 above)
Comment (CaerusKaru) [reply src/lib/media-query/match-media.ts]: Added, great suggestion!

File: src/lib/server/server-provider.ts (hunk 5 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Fixed

File: src/lib/server/server-provider.ts (hunk 6 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Do you have a CSS formatter in mind? Maybe this can be done in a follow-up PR?

File: src/lib/server/server-provider.ts (hunk 7 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

File: src/lib/server/server-provider.ts (hunk 10 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Removed

File: src/lib/server/server-provider.ts (hunk 13 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

File: src/lib/utils/styling/browser-provider.ts (hunk 16 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/browser-provider.ts]: Removed

File: src/lib/utils/styling/styler.ts (hunk 22 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/styler.ts]: This is legacy code, removed. (The idea was that when calling a function, you could call it with string and value individually instead of providing an object)

File: src/lib/utils/styling/styler.ts (hunk 21 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/styler.ts]: Now that the `MatchMedia` has been changed (see above), the only three places this still happens are the following: ImgSrc, ShowHide, and Styler. In each case, the check only changes about two lines of code in very critical ways. I think it would duplicate a lot of code to change and as it stands now doesn't benefit much to factor out.

File: src/lib/utils/styling/styler.ts (hunk 20 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/styler.ts]: Changed to `NOTE`

File: src/lib/utils/styling/server-token.ts (hunk 18 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/server-token.ts]: Added

File: src/lib/utils/styling/server-stylesheet.ts (hunk 17 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/server-stylesheet.ts]: Added

File: src/lib/utils/styling/styler.ts (hunk 19 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/styler.ts]: I vaguely recall @ThomasBurleson saying the same. The original name was `StyleUtils`, I'll change it back.

File: src/lib/utils/styling/browser-provider.ts (hunk 14 above)
Comment (CaerusKaru) [reply src/lib/utils/styling/browser-provider.ts]: Added

File: src/lib/server/server-provider.ts (hunk 8 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Removed from above comment

File: src/lib/server/server-provider.ts (hunk 12 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

File: src/lib/server/server-provider.ts (hunk 11 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

File: src/lib/server/server-provider.ts (hunk 9 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

File: src/lib/server/server-provider.ts (hunk 5 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Added

**Comment by CaerusKaru** [comment]
@jelbourn Comments addressed, PTAL

File: src/lib/media-query/match-media-provider.ts [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {
+  NgZone,
+  PLATFORM_ID,
+  InjectionToken,    // tslint:disable-line:no-unused-variable
```
Comment (jelbourn) [inline src/lib/media-query/match-media-provider.ts]: You can avoid suppressing the lint error by explicitly using `InjectionToken` in the place where it is inferred

File: src/lib/media-query/match-media-provider.ts [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+import {DOCUMENT, isPlatformServer} from '@angular/common';
+
+import {MatchMedia} from './match-media';
+import {ServerMatchMedia} from './server-match-media';
+
+/**
+ * Create a version of MatchMedia compatible with the current
+ * platform
+ */
+export function MEDIA_QUERY_LIST_FACTORY(platformId: Object,
```
Comment (jelbourn) [inline src/lib/media-query/match-media-provider.ts]: Instead of having one provider that returns either service, there should just be two different providers. The appropriate provider is then included in each of the browser / server module.

File: src/lib/media-query/server-match-media.ts [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+import {MatchMedia} from './match-media';
+
+/**
+ * Special server-only class to simulate a MediaQueryList and
+ * - supports manual activation to simulate mediaQuery matching
+ * - manages listeners
+ */
+export class ServerMediaQueryList implements MediaQueryList {
+  private _isActive = false;
+  private _listeners: Array<MediaQueryListListener> = [];
```
Comment (jelbourn) [inline src/lib/media-query/server-match-media.ts]: `MediaQueryListListener[]`

File: src/lib/media-query/server-match-media.ts [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+   * listeners and clearing the internal list
+   */
+  destroy() {
+    this.deactivate();
+    this._listeners = [];
+  }
+
+  /**
+   * Notify all listeners that 'matches === TRUE'
+   */
```
Comment (jelbourn) [inline src/lib/media-query/server-match-media.ts]: You can collapse short comment to a single line:
```ts
/** Notify all listeners that 'matches === TRUE' */
```

File: src/lib/server/server-provider.ts [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+  BreakPoint,
+  BREAKPOINTS,
+  CLASS_NAME,
+  MatchMedia,
+  ServerStylesheet,
+  SERVER_TOKEN,
+  ServerMatchMedia
+} from '@angular/flex-layout';
+
+let UNIQUE_CLASS = 0;
```
Comment (jelbourn) [inline src/lib/server/server-provider.ts]: `UNIQUE_CLASS` shouldn't be caps case because it's not constant. I typically call this `nextId` in material ([example](https://github.com/angular/material2/blob/3aceb7361cc34ad987f7b1ca39339d3203248341/src/cdk/accordion/accordion-item.ts#L24))

File: src/lib/utils/styling/browser-provider.ts (hunk 15 above)
Comment (jelbourn) [reply src/lib/utils/styling/browser-provider.ts]: You can avoid the suppression by explicitly using them in the code where they are now inferred.

File: src/lib/media-query/match-media-provider.ts (hunk 25 above)
Comment (CaerusKaru) [reply src/lib/media-query/match-media-provider.ts]: Fixed

File: src/lib/media-query/server-match-media.ts (hunk 26 above)
Comment (CaerusKaru) [reply src/lib/media-query/server-match-media.ts]: Fixed (and fixed the source for this)

File: src/lib/media-query/server-match-media.ts (hunk 27 above)
Comment (CaerusKaru) [reply src/lib/media-query/server-match-media.ts]: Fixed

File: src/lib/server/server-provider.ts (hunk 28 above)
Comment (CaerusKaru) [reply src/lib/server/server-provider.ts]: Fixed

**Comment by CaerusKaru** [comment]
Comments addressed again, PTAL

**Comment by ThomasBurleson** [comment]
Merged with SHA cf5266adc71cb377a54020d64a500d5f789450ed
